src/hotspot/share/compiler/oopMap.cpp

```diff
@@ -229,12 +235,17 @@ static void trace_codeblob_maps(const frame *fr, const RegisterMap *reg_map) {
 
 void OopMapSet::oops_do(const frame *fr, const RegisterMap* reg_map, OopClosure* f) {
   // add derived oops to a table
-  all_do(fr, reg_map, f, add_derived_oop, &do_nothing_cl);
+  if (!StackWatermarkSet::has_watermark(reg_map->thread(), StackWatermarkKind::gc) ||
+      DerivedPointerTable::is_active()) {
+    all_do(fr, reg_map, f, add_derived_oop, &do_nothing_cl);
+  } else {
+    all_do(fr, reg_map, f, process_derived_oop, &do_nothing_cl);
+  }
```

Comment:
I wonder if we shouldn't hide the StackWatermarkSet in the GC code, and not "activate" the DerivedPointerTable when a SWS is used? Isn't it the case that already don't enable the table for ZGC? Couldn't this simply be:
`
  if (DerivedPointerTable::is_active()) {
    all_do(fr, reg_map, f, add_derived_oop, &do_nothing_cl);
  } else {
    all_do(fr, reg_map, f, process_derived_oop, &do_nothing_cl);
  }
`

Reply:
The problem isn't the GC code deciding to use or not use the derived pointer "table". It's the shared code uses of it that is the problem here. The table is explicitly activated by shared code, when for example using the JFR leak profiler, JVMTI heap walks, etc. This code makes the selection a GC choice when invoked by GC code, and a shared runtime choice when invoked by the shared runtime.

Reply:
After discussions off-line, I think passing in an explicit derived pointer iteration mode into oops_do, is much better than trying to figure out who the caller is from this context.